source/language/types.rst

```diff
@@ -373,3 +375,126 @@ variables whose values may only be known at run time.
++--------------+------------+------------+------------+-------------+------------+------------+------------+------------+
+| **angle**    | :gbg:`Yes` | :rbg:`No`  | :rbg:`No`  | :rbg:`No`   | :center:`-`| :gbg:`Yes` | :rbg:`No`  | :rbg:`No`  |
++--------------+------------+------------+------------+-------------+------------+------------+------------+------------+
+| **bit**      | :gbg:`Yes` | :gbg:`Yes` | :gbg:`Yes` | :rbg:`No`   | :gbg:`Yes` | :center:`-`| :rbg:`No`  | :rbg:`No`  |
++--------------+------------+------------+------------+-------------+------------+------------+------------+------------+
+| **duration** | :rbg:`No`  | :rbg:`No`  | :rbg:`No`  | :rbg:`No*`  | :rbg:`No`  | :rbg:`No`  | :center:`-`| :rbg:`No`  |
++--------------+------------+------------+------------+-------------+------------+------------+------------+------------+
+| **qubit**    | :rbg:`No`  | :rbg:`No`  | :rbg:`No`  | :rbg:`No`   | :rbg:`No`  | :rbg:`No`  | :rbg:`No`  | :center:`-`|
++--------------+------------+------------+------------+-------------+------------+------------+------------+------------+
+
+\*Note: ``duration`` values can be converted to ``float`` using the division operator. See :ref:`divideDuration`
+
+Casting from bool
+~~~~~~~~~~~~~~~~~
+
+``bool`` values cast from ``false`` to ``0.0`` and from ``true`` to ``1.0`` or 
+an equivalent representation. ``bool`` values can only be cast to ``bit[1]`` 
+(a single bit), so explicit index syntax must be given if the target ``bit``
+has more than 1 bit of precision.
+
+Casting from int/uint
+~~~~~~~~~~~~~~~~~~~~~
+
+``int[n]`` and ``uint[n]`` values cast to the standard types mimicking C99
+behavior. Casting to ``bool`` values follows the convention ``val != 0``.
+As noted above, if the value is too large to be represented in the
+target type the result is implementation-specific. Casting to ``bit[m]`` 
+is only allowed when ``m==n``. If the target 
+``bit`` has more or less precision, then explicit slicing syntax must be given.
+As noted, the conversion is done assuming a little-endian 2's complement 
+representation.
+
+Casting from float
+~~~~~~~~~~~~~~~~~~
+
+``float[n]`` values cast to the standard types mimicking C99 behavior (*e.g.* 
+discarding the fractional part for integer-type targets). As noted above, 
+if the value is too large to be represented in the 
+target type the result is implementation-specific. 
+Casting a ``float`` value to an ``angle[m]`` is accomplished by first 
+performing a modulo 2π operation on the float value. The resulting value 
+is then converted to the nearest ``angle[m]`` possible. In the event of a 
+tie between two possible nearest values the result is the one with an even 
+least significant bit (*i.e.* round to nearest, ties to even).
+
+Casting from angle
+~~~~~~~~~~~~~~~~~~
+
+``angle[n]`` values cast to ``bool`` using the convention ``val != 0.0``. 
```

Comment:
I'm not 100% sure that this provides enough utility to justify a casting rule, when `val != 0` should always work.  Maybe it's just me being nervous about the arithmetic on a ring again, though, so I'm happy to leave it if others think it's worthwhile.  Mostly it's that I'm so conditioned to expect that `bool(2 * x) == bool(x)` for _all_ `x`, but `angle` breaks that.  It's not the fault of the casting rule, though - the `val != 0` form also breaks that equality, but it just feels more explicit.

Possibly best leave it since it was already discussed, and just let me keep being nervous about the `angle` type to myself.

Reply:
I don't think you have to be nervous about ``angleVal != 0`` for angle values; this corresponds to something thoroughly unambiguous to me at least (albeit the ``angle`` value of 0 is not identical to the real number 0). However, I think that from a legibility perspective, it would be preferable for the programmer to write ``angleVal != 0`` rather than ``(bool)angleVal``; so for completely independent reasons I would be happy with disallowing casts from ``angle`` to ``bool``.

Reply:
Yeah, I'm not nervous about the actual meaning of that cast - `val != 0` is definitely unambiguous, and it does produce the results I'd expect, given any one single angle.  I think it's just that all my tools for reasoning about computer descriptions of numeric types is broken when I have to think about arithmetic on a ring as a feature, not "undefined behaviour".